Add proposal for breaking changes (#37779)
dotnet/runtime: docs/project/breaking-change-process.md

> # Breaking change process

Reviewer: Not all breaking changes are created equal.  Removing support for some feature is likely much more impactful than, say, switching the order of two argument validation exceptions that might occur if multiple invalid inputs are passed to a method.  Practically any bug fix can break someone.

What is the bar used for this process to define "breaking change"?

Author: I have a few thoughts on this:

- It's hard to define a definite bar. 
- The breaking change docs that I linked to were the ones that were in the repo, mostly from the old corefx repo, I believe. We can file issues if we think they are inadequate.
- In terms of determination, I think of it as a layered approach. People should make a first determination in the early design phase. It would be late, but the person pressing the "big green button" needs to make a determination of the information available on whether a change is breaking. The last resort is .NET users coming to this conclusion in preview or stable releases.

In short, this document doesn't define that, and I think that's fine.

> Breaking changes make it difficult for people to adopt new versions of .NET. We should avoid making them. In the rare case that you believe a breaking change is unavoidable, you should follow the process outlined in this document. The intent of the process is to get feedback. This might result in previously unseen  opportunities to achieve your goals without breaks, clever ways to limit breaks to more niche scenarios, and (most importantly) better understand the scope of impact to .NET users.

Reviewer: ```suggestion
Breaking changes make it difficult for people to adopt new versions of .NET. We strive to avoid making them. In the rare case when a breaking change is unavoidable or the best option when all factors are considered, you should follow the process outlined in this document. The intent of the process is to get feedback. This might result in previously unseen  opportunities to achieve your goals without breaks, clever ways to limit breaks to more niche scenarios, and (most importantly) better understand the scope of impact to .NET users.
```

This process does not result in approval or denial of a breaking change request. It results in feedback, which might be varied or uniform. In the best case, the feedback is a call to action to you, that provides a direction on how you should proceed. The people that are responsible for reviewing your PR are expected to use the information collected by this process as part of their determination of whether your change is appropriate to merge.

> You should start this process as soon as possible, preferably before any code is written, in the design phase. The breaking change issue should be paired with a design document that helps reviewers understand the broader scope, goals and constraints of your change. If you don't yet have a design document, you are encouraged to first write such a document, to enable people giving feedback to develop a deeper level of understanding on your approach. If you do not do this, you might stall the process.

Reviewer: > You should start this process as soon as possible, preferably before any code is written, in the design phase.

For smaller efforts, those are often one in the same, e.g. you don't discover that there needs to be a change in behavior until you're almost done implementing whatever it is you're implementing.

> You should start this process as soon as possible, preferably before any code is written, in the design phase. The breaking change issue should be paired with a design document that helps reviewers understand the broader scope, goals and constraints of your change. If you don't yet have a design document, you are encouraged to first write such a document, to enable people giving feedback to develop a deeper level of understanding on your approach. If you do not do this, you might stall the process.

Reviewer: > The breaking change issue should be paired with a design document that helps reviewers understand the broader scope, goals and constraints of your change.

Wouldn't the issue itself just contain all of those details?  Why is a separate "document" needed?

There are people on the team that can help you work through a breaking change, if you want to engage with them before starting this more formal process. They are part of the [dotnet/compat team](https://github.com/orgs/dotnet/teams/compat) . Feel free to @mention them on issues to start a dialogue.

> ## Process

Reviewer 2: Are we supposed to create new issues for issues that track design work in case this design has been identified as potentially being breaking or is it OK if we apply this label to such issues?

Author: That one is easy. You should create an issue when you start wanting feedback (not solely on compatibility).

> 1. Create an issue that describes the breaking change

Reviewer: A common workflow is there's an issue opened to describe a feature, the feature is discussed, and then subsequently a PR is opened to implement that feature.

Is this process proposing yet another issue be opened to describe potential breaks in the first issue?

   * Mark the issue with the [breaking-change](https://github.com/dotnet/runtime/labels/breaking-change) label
   * Goals and motivation for the change
   * Pre-change behavior
   * Post-change behavior
   * Errors or other behavior you can expect when running old code that breaks

>    * Workarounds and mitigations, including [AppContext switches](https://docs.microsoft.com/dotnet/api/system.appcontext)

Reviewer 3: Do we have guidance somewhere indicating when a change justifies adding a switch? My impression was that our bar was high for this, to avoid reducing maintainability. 

Currently I see a couple for globalization invariant mode, a couple for threadpool, 3 for HTTP/2, a couple others. So it seems we use them for 1-2% of breaking changes.

Also we can presumably say that the switch would always be to switch off the default behavior rather than vice versa.

Author: I am happy to remove this. I saw somewhat similar text in doc Priya gave me, and was wanting to pull some of thinking from it. I look to you for guidance on this. I'm fully neutral on it.

Reviewer 2: I think instead of removing this I think it would be much better if your doc could start talking about what criteria we use to decide whether we do an app compat switch or not.

Reviewer 4: >  talk about what criteria we use

Here is one: https://github.com/dotnet/runtime/issues/8655#issuecomment-415103199

   * Link to the issue for the feature or bug fix that the breaking change is associated with.
   * Reference this issue from associated PRs.
2. Share your issue with whomever you see as stakeholders
   * Please @mention [dotnet/compat team](https://github.com/orgs/dotnet/teams/compat) team
   * Engage with people commenting on the issue, with the goal of deriving the most feedback on your proposed breaking change
   * This may involve a lot of explanation on your part. A well-written design doc can ease the burden here.
3. Mark associated PRs with the [breaking-change](https://github.com/dotnet/runtime/labels/breaking-change) label, and link to your breaking change issue.
4. Once the PR is merged, create a [docs issue](https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md).

>    * Clarify which .NET preview the break will be first released in.

Reviewer 2: Should we add an announcement in dotnet/announcement too? Presumably this makes sense for some, but not all. How do we decide?

Author: I think this would be a rare case.

## Examples

The following are good examples of breaking change issues

* https://github.com/dotnet/runtime/issues/28788
* https://github.com/dotnet/runtime/issues/37672

> 

Reviewer 2: nit: should probably be a single line

# Breaking change process

> Breaking changes make it difficult for people to adopt new versions of .NET. We strive to avoid making them. In the rare case when a breaking change is unavoidable or the best option when all factors are considered, you should follow the process outlined in this document. The intent of the process is to get feedback. This might result in previously unseen  opportunities to achieve your goals without breaks, clever ways to limit breaks to more niche scenarios, and (most importantly) better understand the scope of impact to .NET users.

Reviewer 4: ```suggestion
Breaking changes make it difficult for people to adopt new versions of .NET. We strive to avoid making them. In the rare case when a breaking change is unavoidable or the best option when all factors are considered, you should follow the process outlined in this document. The intent of the process is to get feedback. This might result in previously unseen opportunities to achieve your goals without breaks, clever ways to limit breaks to more niche scenarios, and (most importantly) better understand the scope of impact to .NET users.
```

This process does not result in approval or denial of a breaking change request. It results in feedback, which might be varied or uniform. In the best case, the feedback is a call to action to you, that provides a direction on how you should proceed. The people that are responsible for reviewing your PR are expected to use the information collected by this process as part of their determination of whether your change is appropriate to merge.

> You should start this process as soon as possible, preferably before any code is written, in the design phase. This isn't always possible, since the change in behavior might not be discovered until you're almost done building a feature. In the case of a significant breaking change, you  should provide reviewers with a design document so that they can understand the broader scope, goals and constraints of your change.

Reviewer 4: ```suggestion
You should start this process as soon as possible, preferably before any code is written, in the design phase. This isn't always possible, since the change in behavior might not be discovered until you're almost done building a feature. In the case of a significant breaking change, you should provide reviewers with a design document so that they can understand the broader scope, goals and constraints of your change.
```
(extra space)

There are people on the team that can help you work through a breaking change, if you want to engage with them before starting this more formal process. They are part of the [dotnet/compat team](https://github.com/orgs/dotnet/teams/compat) . Feel free to @mention them on issues to start a dialogue.

## Process

1. Create or link to an issue that describes the breaking change, including the following information:
   * Mark with the [breaking-change](https://github.com/dotnet/runtime/labels/breaking-change) label
   * Goals and motivation for the change

>    * Pre-change behavior

Reviewer 3: Is it interesting to note what versions it affects? Eg., this API was added in .NET Core and the break does not affect folks porting code from .NET Framework.

Author: Good call. I added a note on that. This allowed me to bring in relevant context that I'd forgotten about (.NET Framework compatibility mode).

   * Workarounds and mitigations, including [AppContext switches](https://docs.microsoft.com/dotnet/api/system.appcontext)
   * Link to the issue for the feature or bug fix that the breaking change is associated with.
   * Reference this issue from associated PRs.
2. Share your issue with whomever you see as stakeholders
   * Please @mention [dotnet/compat team](https://github.com/orgs/dotnet/teams/compat) team
   * Engage with people commenting on the issue, with the goal of deriving the most feedback on your proposed breaking change
   * This may involve significant explanation on your part. A well-written design doc can ease the burden here.
3. Mark associated PRs with the [breaking-change](https://github.com/dotnet/runtime/labels/breaking-change) label, and link to your breaking change issue.
4. Once the PR is merged, create a [docs issue](https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md).
   * Clarify which .NET preview the break will be first released in.
5. Breaking change issues can be closed at any time after the PR is merged. Best practice is waiting until the change has been released in a public preview.

Notes:

> * [Guidance for when to use AppContext switched](https://github.com/dotnet/runtime/issues/8655#issuecomment-415103199)

Reviewer 3: ```suggestion
* [Guidance for when to use AppContext switches](https://github.com/dotnet/runtime/issues/8655#issuecomment-415103199)
```
